ikarus/utils/observer/observer.hh

```diff
@@ -13,35 +13,57 @@ template <typename MessageType>
+
   void update(MessageType message) {
-    assert(MessageType::END != message && "The END enum type should not be used");
-    assert(MessageType::BEGIN != message && "The BEGIN enum type should not be used");
+    checkMessageType(message);
     updateImpl(message);
   }
+
   void update(MessageType message, double val) {
-    assert(MessageType::END != message && "The END enum type should not be used");
-    assert(MessageType::BEGIN != message && "The BEGIN enum type should not be used");
+    checkMessageType(message);
     updateImpl(message, val);
   }
 
-  void update(MessageType message, int intVal, double val1, double val2) {
-    assert(MessageType::END != message && "The END enum type should not be used");
-    assert(MessageType::BEGIN != message && "The BEGIN enum type should not be used");
-    updateImpl(message, intVal, val1, val2);
-  }
+  void update(MessageType message, int val) {
+    checkMessageType(message);
+    updateImpl(message, val);
+  };
+
+  void update(MessageType message, std::string val) {
+    checkMessageType(message);
+    updateImpl(message, val);
+  };
+
+  void update(MessageType message, int val1, double val2) {
+    checkMessageType(message);
+    updateImpl(message, val1, val2);
+  };
+
+  void update(MessageType message, int val1, double val2, std::string val3) {
+    checkMessageType(message);
+    updateImpl(message, val1, val2, val3);
+  };
 
   void update(MessageType message, const Eigen::VectorXd& vec) {
-    assert(MessageType::END != message && "The END enum type should not be used");
-    assert(MessageType::BEGIN != message && "The BEGIN enum type should not be used");
+    checkMessageType(message);
     updateImpl(message, vec);
   }
 
 protected:
-  virtual void updateImpl([[maybe_unused]] MessageType message) {}
-  virtual void updateImpl([[maybe_unused]] MessageType message, [[maybe_unused]] double val) {}
-  virtual void updateImpl([[maybe_unused]] MessageType message, [[maybe_unused]] int intVal,
-                          [[maybe_unused]] double val1, [[maybe_unused]] double val2) {}
+  virtual void updateImpl([[maybe_unused]] MessageType message){};
+  virtual void updateImpl([[maybe_unused]] MessageType message, [[maybe_unused]] double val){};
+  virtual void updateImpl([[maybe_unused]] MessageType message, [[maybe_unused]] int val){};
+  virtual void updateImpl([[maybe_unused]] MessageType message, [[maybe_unused]] std::string val){};
+  virtual void updateImpl([[maybe_unused]] MessageType message, [[maybe_unused]] int val1, double val2,
+                          std::string val3){};
+  virtual void updateImpl([[maybe_unused]] MessageType message, [[maybe_unused]] int val1, double val2){};
   virtual void updateImpl([[maybe_unused]] MessageType message, [[maybe_unused]] const Eigen::VectorXd& vec) {}
+
+private:
+  void checkMessageType(MessageType message) {
+    if (MessageType::END == message) DUNE_THROW(Dune::NotImplemented, "The END enum type should not be used");
```

Comment:
`Dune::InvalidStateException`is more appropriate here, since `NotImplemented` does not make sense.